### PR TITLE
[hold][rfr][feature] Add Veer registration schema

### PR DESCRIPTION
**Pull request by samchrisinger**
# Purpose

From: https://openscience.atlassian.net/browse/OSF-5894

This is a fairly literal interpretation of the contributor's schema. Probably some cleanup is needed, simple proof of concept.

### CODE REVIEW

File: website/project/metadata/schemas.py [hunk 1]
```diff
@@ -38,5 +39,5 @@ def from_json(fname):
     'OSF-Standard Pre-Data Collection Registration',
```
Comment (samchrisinger) [inline website/project/metadata/schemas.py]: We should find a better way to do this.

Comment (JeffSpies) [inline website/project/metadata/schemas.py]: People will start building registries on the OSF a la pre-reg but via their own interfaces. We can re-architect to make that really easy--perhaps with the inclusion of JamDB buckets on GUIDs.

**Comment by AnnaVantVeer** [comment]
These things are the first that come to mind:
- the overview on the right of the page should have an A, B, C structure and B is missing, I see where this went wrong: there were headings and subheadings (but I did not know how to make subheadings, see comment in original markdown for which title should be subheading). Is it possible to have the main headings (like the first two that start with A, which are correctly displayed on the right), and also have a subheading above the questions on that specific page (which is not in the overview on the right)?
- all items have "(optional)" behind them, can we simply delete that text? Although it should be possible to proceed with the registration if you don't have an answer to one question, we do not want to give users the idea that things are optional. In fact, I wouldn't mind if all items are 'forced response', meaning something should be filled out (even if it is a short as saying "NA")
- question 4a (under "A. hypothesis, recommended elements") says that a picture would be advised, meaning if it is possible to code this it would be very cool if a user can upload a picture there. If not, then this question needs to be rewritten.

I will upload a Word doc from which I made the code, this will probably clarify the structure that is intended. Thanks so much!
[Pre-Registration in Social Psychology - Template according to van t Veer and Giner-Sorolla 2016.docx](https://github.com/CenterForOpenScience/osf.io/files/202633/Pre-Registration.in.Social.Psychology.-.Template.according.to.van.t.Veer.and.Giner-Sorolla.2016.docx)

**Comment by samchrisinger** [comment]
@AnnaVantVeer can you please take a look at the updated form?

screen shot 2016-04-05 at 11 41 42 am
screen shot 2016-04-05 at 11 41 58 am
screen shot 2016-04-05 at 11 42 11 am
screen shot 2016-04-05 at 11 42 20 am
screen shot 2016-04-05 at 11 42 32 am
screen shot 2016-04-05 at 11 42 45 am
screen shot 2016-04-05 at 11 43 00 am
screen shot 2016-04-05 at 11 43 12 am
screen shot 2016-04-05 at 11 43 22 am

**Comment by evomellor** [comment]
It looks good, I think the organization is as expected.

I like the array of multiple analyses, I think that will be a nice feature for users to organize their plans. Does that appear as standard? In other words, if a user only has one main prediction, will they be required to enter "N/A" for all of the subsequent fields? If so, I think it may make sense to make every field from the "second main prediction" onward optional for submission. Unfortunately, it will further muddle the purpose of required/optional for the form instead of the registration's validity, but I think it could result in unnecessary annoyance if all of those fields are required.

**Comment by icereval** [comment]
@evomellor, @samchrisinger, let me know when the details are ironed out, seems to be some discrepancy...

**Comment by AnnaVantVeer** [comment]
This looks great, thanks Sam! I agree with David about that it would be best to make any additional analysis plan for the hypotheses (other than the first one) optional. If people only have one hypothesis, they won't have to put NA in all the other boxes. I guess the website does not support a function that would ask users upfront to say how many analysis plans for hypotheses they want to insert, and then it adapts to this. right? So this would be a good alternative (making the first required and the others optional), it just takes some extra scrolling.
How does the scrolling feel right now within one of the three main pages? It the pages are very long, we could make 6: for each one we have now, split it into 'Essential elements' and 'Recommended elements'. Because in this version the numbering is gone (which is fine, by the way) we can now do this more elegantly then before (because number 9 (under exclusion criteria) was recommended even though it seemed structured under essential, this reflects my bad way of making a table in Word). If we do it in this 6 pages way, in the wording for number 9 we can delete "(recommend)" and this would be the only question on that "Method - Recommended elements" page. Is this reasonably easy to do Sam?

**Comment by samchrisinger** [comment]
@AnnaVantVeer I went ahead and made the second, third, fourth, and further analysis plans options. Regarding how the amount of scrolling feels-- would you be willing to create a [https://staging.osf.io/](https://staging.osf.io/) account and test this out when we put this on our staging site? But switching to 6 pages is not too hard either way. 

We'd discussed letting users enter an arbitrary number of analysis plans (up to 10, for example), but decided we didn't have a strong enough use case to merit implementing right away. @evomellor, maybe we should re-open discussion on this?

**Comment by evomellor** [comment]
Thanks @samchrisinger 
For this schema, I think some visual separation between each of the multiple analyses will be necessary (e.g. a white space between "if using techniques other than null..." and "Describe the analyses that will test...") I see you put in an extra grey line, but's is still hard to the see that these are separate fields. Let's wait for @AnnaVantVeer feedback before making that change, though.

Yes, I think it is time to consider making multiple hypotheses and multiple analyses more feasible for the competition. I have the draft of a proposal for that and will ping you for discussion about that sometime in the next few days.

**Comment by AnnaVantVeer** [comment]
Hi all!
I've started to make a staging account (I presume the staging account is different from my normal osf account) - waiting for the registration email as we speak.
I'm incorporating some comments here from Roger (co-author of the paper) here:
1) question about the text boxes: they might seem small, and although users can widen them with the lower right corner, is it at al possible to start out with bigger ones for a couple of areas? (like the hypothesis one, because there people might want to type a lot)
2) Over skype we'd discussed the instructions that come at the beginning (and the NA's are explained there beautifully, thanks) -  would it be an idea to also put our osf project page there as a reference for where people should go if they are confused etc? It could be something like this: first comes the N/A instruction "For ..N/A.. field. See van 't Veer & Giner-Sorolla (2016) or https://osf.io/56g8e/ for additional information."
3) coming back to David's suggestion about the grey space between separate analyses, yes that sounds like a great idea to have a visually clear separation between them. Thanks for thinking along this way! An thanks for making the 2nd 3rd etc optional, Sam, that saves a lot of N/A's.
3a) If you open the discussion about users inserting an arbitrary number for their hypotheses (somewhere in the future) I presume we could still implement that at a later stage? Or should we wait for this function so that we don't need any other changes once we go live?
4) I'll be testing the feel of the scrolling in order to come back to you about the open topic of whether we make the list on the left into a 6 item list (I have a slight preference for this, because then people can see the step structure a little more; you might be discouraged if you see a long list of items to fill out under A, only to find that all the things at the end of this list were optional). 
I'll be checking out how this staging account works, no registration confirmation email yet ... (sorry for the delay in response by the way, this is what happens when working on opposite sites of the ocean, but I'll to be available tonight too)

**Comment by samchrisinger** [comment]
> question about the text boxes: they might seem small, and although users can widen them with the lower right corner, is it at al possible to start out with bigger ones for a couple of areas? 

Yes this is possible. Please let me know which fields need larger textareas. 

> would it be an idea to also put our osf project page there as a reference for where people should go if they are confused etc? It could be something like this: first comes the N/A instruction "For ..N/A.. field. See van 't Veer & Giner-Sorolla (2016) or https://osf.io/56g8e/ for additional information."

We can absolutely do this. Just let me know if you'd like this to be included. 

> coming back to David's suggestion about the grey space between separate analyses, yes that sounds like a great idea to have a visually clear separation between them. 

I tried doing:
screen shot 2016-04-08 at 10 49 10 am

Notice "Second Analysis" in a separate box. The other labels are "Third ...", "Fourth ...", "Further Analyses". Does this work? 

> If you open the discussion about users inserting an arbitrary number for their hypotheses (somewhere in the future) I presume we could still implement that at a later stage? Or should we wait for this function so that we don't need any other changes once we go live?

Yes, I think this would be a backwards compatible change.

**Comment by AnnaVantVeer** [comment]
ok great.
re your last two points: great.
re second point, yes please include that extra line in the instructions.
re textarea's:

Text boxes that could be larger (e.g., four times the size now);
- first field under hypotheses: ’Describe the (numbered) hypotheses…
- second field under methods: ‘Dependent variables’
- last field under methods: “describe all manipulations, measures, …”

Text boxes that could be larger (e.g., twice the size now);
- ‘For original research, add rationales ..’
- first field under methods: ’Design […] a) whether they are within […] b) the relationship between then (..).
- under analysis plan: only for the first hypothesis analysis: “the relevant variables and how they are calculated;”

(if there is only one larger size then all the above:)

File upload:
- upload visual area for the file is large now, but maybe that srinks according to the file, I’ll check this when my account is set up (just to be sure, we are uploading a file from the computer, not from files on your osf project page, right? don't answer this if I can find out when I do the pilotting to save you time:) )
- could we add a file upload under ‘justify planned sample’? Then this would be an additional item (so next to the text area for ‘justify planned sample’) with this text: 
“If applicable, you can upload a file related to your power analysis here (e.g., a protocol of power analyses from G*Power, a script, a screenshot, etc.).”

I'll likely test via the staging account tomorrow, and since I don't assume you guys work on the weekend I'll let you know if I come across anything but we'll tackle the rest next week. If the above can be tackled before I test that would be great though.
Thanks again!

**Comment by AnnaVantVeer** [comment]
Okay! I've finally made my staging account, and next to the above edits I have some suggestions (I number them because that may make it easier to respond):
1) if you can make it into six pages instead of 3 that would be great, I think it would be a little bit more elegant. The list on the left then becomes: A. Hypotheses - Essential elements, Recommend elements, B. Methods - Essential elements, Recommended elements, C. Analysis plan - Essential elements, Recommended elements.
1a) the titles that are on the pages now already can stay as they are (some are called 'Description of essential elements (required)' some are called 'Confirmatory analyses (required)'.
2) making 6 pages means that for the methods part we need to decide what to do in terms of 'splitting' stuff into two (in order to have one Essential page, and one Recommended page). Could you do the following? Make the question that reads '(Recommended element) Set fail-safe levels of exclusion at which the whole study needs to be stopped, altered, and restarted.' on the Recommended elements page? It would need a subtitle (like it has now) their too, namely 'Procedure' (under the title that is in bold 'Recommended elements (optional)'). 
2a) it now seems a little silly to have only one question on this page. This got me thinking. If it is not too much trouble, I would love if you can add a file-uplaod-question here, like so: "If applicable, you can upload any files related to your methods and procedure here (e.g., a paper describing a scale you are using, experimenter instructions, etc.)"
3) I'm sorry I keep asking for more, but if the above is doable, then please also add this to the Recommended elements of the C. Analysis plan:
"Optionally, upload any files here that are related to your analyses (e.g., syntaxes, scripts, etc.)."

4) I was missing some questions that were in the coding before, but maybe you've removed them because they look a lot like the standard osf form. Is it possible to insert them at the end of the form, meaning there will an additional 7th page that is required? These are the questions:
        "questions": [{
            "qid": "datacompletion",
            "title": "Has data collection begun for this project?",
            "nav": "Data Completion",
            "type": "choose",
            "format": "singleselect",
            "options": ["No, data collection has not begun", "Yes, data collection is underway or complete"],
            "description": "Please choose"
        },{
            "qid": "looked",
            "title": "Have you looked at the data?",
            "nav": "Looked at Data",
            "type": "choose",
            "format": "singleselect",
            "options": ["Yes", "No"],
            "description": "Please choose"
        },{
            "qid": "dataCollectionDates",
            "title": "The (estimated) start and end dates for this project are:",
            "nav": "Data collection period",
            "type": "string",
            "format": "textarea"
        },{
            "qid": "additionalComments",
            "title": "Any additional comments before I pre-register this project:",
            "nav": "Additional comments",
            "type": "string",
            "format": "textarea"
        }]

Sorry if this causes extra work, I hope it's still doable. If there are any questions/issues/etc we could skype. I think the remaining issues are in the last two comments rounds (including your suggestion for making the second/third/fourth analyses optional like in your screenshot). Hope everything is clear, let me know otherwise. Thanks so much again, Anna

**Comment by samchrisinger** [comment]
@AnnaVantVeer doing a little more work on this today...

I've:
-  increased the sizes of the requested textareas
- added a file upload after ‘justify planned sample’
- split section B into two pages
- added the other requested file uploads

> I was missing some questions that were in the coding before, but maybe you've removed them because they look a lot like the standard osf form. 

My goal was to match [the attached schema](https://github.com/CenterForOpenScience/osf.io/files/202633/Pre-Registration.in.Social.Psychology.-.Template.according.to.van.t.Veer.and.Giner-Sorolla.2016.docx) as faithfully as possible. Before I add the extra page can you please answer a few questions: 

- How should the page appear on the left navbar? I.e. what should it be titled? 
screen shot 2016-04-12 at 3 29 43 pm
- Are any of these questions required? Optional?
- Do you have expectations about what size the textareas should be for any of the new items?

Please let me know if I've missed anything.

**Comment by samchrisinger** [comment]
Sorry, to be clear I've made these changes locally, but they are not yet live on staging.

As soon as we can resolve the questions above we can push this out to the staging site.

**Comment by AnnaVantVeer** [comment]
Hi Sam, great thanks. And sorry yes I did not specify all things that you need. Answering your questions in a row:
- left navbar: the last page can be called 'Final questions'. The other sidebar items look good, but can you skip the repeating of 'A hypothesis' etc for the recommended pages? I think it will be clear that they fall under these A B C headings and this will save some repetition. I'll copy them here again just to be sure 
there is no confusion;  A. Hypotheses - Essential elements, Recommend elements, B. Methods - Essential elements, Recommended elements, C. Analysis plan - Essential elements, Recommended elements.
please note the screenshot above has some different spelling of the word Recommended.
- first two questions on the last page required, last two questions optional (if possible, this can be said right next to them, like you have it in other forms where the word is behind the question: "(optional)".
- text areas of new items on this last page can be small (the size we started with for all questions).
Does this answer all questions? I will check again late tonight (my time) because this will make sure you can go ahead if you have any more questions, this timezone lag is not optimal :)

**Comment by samchrisinger** [comment]
Yes, thanks for the answers!

**Comment by AnnaVantVeer** [comment]
Hi Sam, did any new questions come up about the above? Cheers, Anna

**Comment by samchrisinger** [comment]
@AnnaVantVeer, can you please check the updated schema on https://staging.osf.io/?

**Comment by AnnaVantVeer** [comment]
Looks good!
Couple of small things:
- could you name the 'second analysis' etc, 'prediction' instead of analysis? So Second prediction, third prediction, 'further predictions' etc. Sorry I did not notice this before (seeing it in the form made me realise that people can have multiple analyses to test their first main prediction, their second etc.
- file upload: if one file is uploaded, I cannot find another way to add more files. As in the attached screenshot, I only see the option to download it again. Am I doing something wrong maybe? 
- is it possible for Roger to get a staging account as well? if yes, I will tell him to get an account and check out our result (but I am not sure whether just anyone can just make an account...?)
screen shot 2016-04-19 at 19 42 15

I will check everything thoroughly again tomorrow (in daytime for me :) but it looks very good from the looks of it now. Thanks so much Sam!

**Comment by AnnaVantVeer** [comment]
Hi guys, I just finished a registration on the staging site, and I think it looks great! I assume Roger can check it on the staging site too? I'll tell him to have a look, and then I think we can go live!?

**Comment by AnnaVantVeer** [comment]
oh by the way, are you guys aware that the end of the registration form gives an error if you put the embargo more than 4 years ahead, but not the kind of error you expect? see attached
screen shot 2016-04-26 at 09 52 57

**Comment by samchrisinger** [comment]
Hi @AnnaVantVeer.

We've been continuing work on this here: https://github.com/CenterForOpenScience/osf.io/pull/5517. Some extra development was required to support the kind of nesting needed for the new schema.

> could you name the 'second analysis' etc, 'prediction' instead of analysis? So Second prediction, third prediction, 'further predictions' etc. 

This is done.

> file upload: if one file is uploaded, I cannot find another way to add more files. 

We are working on this: 
- https://github.com/CenterForOpenScience/osf.io/pull/5494
- https://github.com/CenterForOpenScience/osf.io/pull/5280

I believe the latter does what you're looking for, but I will need to talk with the author before I can confirm this.

Also, thank you for the heads up regarding the validation for embargo ending date. That is fixed in: https://github.com/CenterForOpenScience/osf.io/pull/5517/commits/7c48db51161a67daba67b2a8cd512818494abd24

**Comment by samchrisinger** [comment]
As a followup, the file picker now support selecting/uploading an arbitrary number of files.

**Comment by AnnaVantVeer** [comment]
Hi Sam, there seems to be some text missing in the template under methods/design: it says:
'Design
 List, based on your hypotheses from section A:
    a) whether they are within- or between-participants
    b) the relationship between them (e.g., orthogonal, nested).'

But the template has this text:
'Design
List, based on your hypotheses from section A:
1.	Independent variables with all their levels
    a.	whether they are within- or between-participants 
    b.	the relationship between them (e.g., orthogonal, nested). 
2.	Dependent variables.
3.	Third variables acting as covariates or moderators.'

Can you fix this?

- When you are on that page, could you also change the wording for the next box (now 'Dependent variables') into: "List dependent variables, or variables in a correlational design"
 
- Under Methods/Recommended elements, under the line 'Set fail-safe levels of exclusion at which the whole study needs to be stopped, altered, and restarted.' can you add a line: "You may pre-determine what proportion of excluded participants will cause the study to be stopped and restarted."

- on the last page, could you change the wording of this line: 'Have you looked at the data?'
into this: "If data collection has begun, have you looked at the data?"

Sorry about these last minute things, they are things that Roger noticed (he finally got on the staging page with my account because he couldn't get it to work before). Hoping you have the time to change these (I've tried to keep them small, textual things). Thanks again Sam!

**Comment by samchrisinger** [comment]
These changes are included in https://github.com/CenterForOpenScience/osf.io/pull/5517/commits/30b2e11d79195e956428988e9c89cc8b12cace0c

**Comment by AnnaVantVeer** [comment]
Hi Sam, great! I checked that link but as you know I am not a code expert; I do not see the first difference in text  on the staging site yet, or is it only on the code page?  In any case, if I read the code right, there is one thing still:

I see in the code that you added: 
"List, based on your hypotheses from section A:\n\n1. Independent variables with all their levels\n    a. whether they are within- or between-participant\n    b. the relationship between them (e.g., orthogonal, nested).\n2. Dependent variables.\n3. Third variables acting as covariates or moderators."

But I think my copying the text in my request above was a little misleading, the only change was for adding the "independent  variables...levels" text, the "a. ... b ..." was already correct, and you don't need to add the last part of this code above (n2. Dependent variables.\n3. Third variables acting as covariates or moderators.) because those are separate boxes below this box, correct? In the separate box 'dependent variables' you did change the wording if I see it correctly.

Let me know if I can check it out on the staging page, and I think we are ready to go live after that! yeey!

**Comment by samchrisinger** [comment]
Thank you for the clarification and for looking over my work.

This now looks like:
screen shot 2016-04-27 at 2 40 59 pm

**Comment by AnnaVantVeer** [comment]
Great Sam!
Anything you need from me to be able to go ahead with putting it online? 
When will the next wave (or what do you call this) be? I have a demo on Friday (15:00 my time), would be very cool if it already works then. 

Cheers!

Anna

**Comment by samchrisinger** [comment]
Hi @AnnaVantVeer if you have time, can you please do a final check of the form on staging?

**Comment by AnnaVantVeer** [comment]
Yes! I checked it! looks good to me.
I see you have some additional text to explain uploading files, and this is not the same for all the spots where people can upload files, but I assume you have a good reason for it.
So go ahead and push this final version through! yeey!

**Comment by AnnaVantVeer** [comment]
Hi Sam, is it correct that I don't see the registration anymore on the staging site as well? then I know not to demo it right now...

**Comment by samchrisinger** [comment]
@AnnaVantVeer that's correct. We're working this morning on getting this back on staging.

**Comment by AnnaVantVeer** [comment]
Hi Sam, any updates on the template? Cheers, Anna

**Comment by sloria** [comment]
@AnnaVantVeer The code for this is currently getting reviewed. Once that's done and we've done QA, we will release it to production. We are hoping to get have it ready early next week.

**Comment by sloria** [comment]
@AnnaVantVeer Just to let you know, the new schema is now live on production.

**Pull request merged**